### PR TITLE
IGNITE-21300 Implement disaster recovery for secondary indexes

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-21300

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuilder.java [hunk 1]
```diff
@@ -85,7 +86,8 @@ class IndexBuilder implements ManuallyCloseable {
      *     <li>Index is built in batches using {@link BuildIndexReplicaRequest}, which are then transformed into {@link BuildIndexCommand}
      *     on the replica, batches are sent sequentially.</li>
      *     <li>It is expected that the index building is triggered by the primary replica.</li>
-     *     <li>If the index has already been built, {@link IndexBuildCompletionListener} will be notified.</li>
+     *     <li>If the index has already been built, {@link IndexBuildCompletionListener#onBuildCompletion} will be notified.</li>
+     *     <li>After index building is complete, {@link IndexBuildCompletionListener#onBuildCompletion} will be notified.</li>
```
Comment (Phillippko) [inline modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuilder.java:90]: onBuildCompletionAfterDisasterRecovery

Comment (Phillippko) [inline modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuilder.java:90]: Maybe rephrase? IMHO now it looks confusing, like a copypaste error (onBuildCompletion in both cases).

File: modules/index/src/test/java/org/apache/ignite/internal/index/IndexBuilderTest.java [hunk 2]
```diff
@@ -149,6 +154,16 @@ void testIndexBuildWithReplicationTimeoutException() {
...
     }
 
+    @Test
+    void testScheduleBuildIndexAfterDisasterRecovery() {
+        CompletableFuture<Void> listenCompletionIndexBuildingAfterDisasterRecoveryFuture =
+                listenCompletionIndexBuildingAfterDisasterRecovery(INDEX_ID, TABLE_ID, PARTITION_ID);
+
+        scheduleBuildIndexAfterDisasterRecovery(INDEX_ID, TABLE_ID, PARTITION_ID, List.of(rowId(PARTITION_ID)));
+
+        assertThat(listenCompletionIndexBuildingAfterDisasterRecoveryFuture, willCompleteSuccessfully());
```
Comment (Phillippko) [inline modules/index/src/test/java/org/apache/ignite/internal/index/IndexBuilderTest.java:164]: Could we also assert that onBuildCompletion method in listener is not triggered? The same for testScheduleBuildIndex, that afterRecovery is not triggered

File: modules/index/src/test/java/org/apache/ignite/internal/index/IndexBuilderTest.java [hunk 3]
```diff
@@ -162,12 +177,43 @@ private void scheduleBuildIndex(int indexId, int tableId, int partitionId, Colle
...
     private CompletableFuture<Void> listenCompletionIndexBuilding(int indexId, int tableId, int partitionId) {
-        CompletableFuture<Void> future = new CompletableFuture<>();
+        var future = new CompletableFuture<Void>();
 
-        indexBuilder.listen((indexId1, tableId1, partitionId1) -> {
-            if (indexId1 == indexId && tableId1 == tableId && partitionId1 == partitionId) {
-                future.complete(null);
+        indexBuilder.listen(new IndexBuildCompletionListener() {
+            @Override
+            public void onBuildCompletion(int indexId1, int tableId1, int partitionId1) {
```
Comment (Phillippko) [inline modules/index/src/test/java/org/apache/ignite/internal/index/IndexBuilderTest.java:198]: Could we also assert that `onBuildCompletionAfterDisasterRecovery`method is not triggered? The same for `listenCompletionIndexBuildingAfterDisasterRecovery`, but ofc `onBuildCompletion`

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuilder.java (hunk 1 above)
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuilder.java:90]: Okay, tell me which option is better so as not to confuse. It's harder for me to rephrase this because I understand how it works.

Comment (Phillippko) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuilder.java:90]: _If the index has already been built or after the building is complete, onBuildCompletion will be notified_?

Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuilder.java:90]: Nice

**Review by Phillippko — APPROVED** [review]
lgtm

**Review by ibessonov — APPROVED** [review]
